extra_geom/detectors.py

```diff
@@ -768,6 +769,123 @@ def offset(self, shift, *, modules=np.s_[:], tiles=np.s_[:]):
         ])
 
 
+class GenericGeometry(DetectorGeometryBase):
+    """A generic detector layout based either on the CrystFEL geom file or on a set of parameters.
+
+    The coordinates used in this class are 3D (x, y, z), and represent metres.
+
+    The :attr:`expected_data_shape` is a following triple :
+
+        1. the number of modules :attr:`n_modules`
+        2. the number of tiles in a module along the slow-scan direction multiplied by
+           the number of slow-scan pixels per tile :attr:`frag_ss_pixels`
+        3. the number of tiles in a module along the fast-scan direction multiplied by
+           the number of fast-scan pixels per tile :attr:`frag_fs_pixels`
+    """
+    detector_type_name = 'Generic Detector'
+
+    def __init__(self, pixel_size: float, fast_pixels: int, slow_pixels: int,
+                 corner_coordinates: [np.ndarray] = [np.zeros(3)],
+                 fs_vec: np.ndarray = np.array([0, 1, 0]),
+                 ss_vec: np.ndarray = np.array([1, 0, 0]),
+                 n_tiles_per_module: int = 1,
+                 tile_gap: float = None,
+                 tile_vec: np.ndarray = None,
+                 ):
+        """ Create a generic detector:
+
+        `pixel_size`: the size of a pixel in meters (reversed CrystFEL's `res`)
+        `fast_pixels`, `slow_pixels`: the size of a tile along fast and slow axes
+        `corner_coordinates`: 3D coordinates of the first pixel of each module
+        `n_modules`: the number of modules, default=1
+        `n_tiles_per_module`:  the number of tiles in each module, default=1
+        `tile_gap`: the gap between two tiles, default=pixel_size
+        `tile_vec`: the direction of tile replication, default=[1, 0, 0]
```

Comment:
I don't find this the easiest thing to follow, but that may be just me. :slightly_smiling_face:

Reply:
I added a bit more to the documentation. 

I used a 3D vector just because there were some already, and that's why I decided that it'd be natural to follow this vector concept with tiles too. If we would decide to use, say, a string variable instead (like `tile_direction = 'left | right | up | down'`) then it required some cumbersome piece of code to interact with numbers.

Reply:
I was thinking of just `'slow_scan'` or `'fast_scan'`, and not allowing the situation where the tile vector is in the opposite direction to one of the pixel vectors, because I don't think that's particularly likely to come up. But if you consider this and still prefer specifying a vector, that's OK with me.

Reply:
Yes, I would stick to a vector. (BTW one can even put tiles on a diagonal with (1, 1, 0) :smiley:)

Reply:
![Your scientists were so preoccupied with whether or not they could, they didn't stop to think if they should](https://user-images.githubusercontent.com/327925/118678399-e497f800-b7f4-11eb-9566-9b3f33c34db4.png)

:wink:

Reply:
:see_no_evil: